socorro/lib/javautil.py

```diff
@@ -102,3 +107,38 @@ def parse_java_stack_trace(text):
             new_exc.additional.append(line)
 
     return new_exc
+
+
+class MalformedJavaException(Exception):
+    pass
+
+
+def validate_java_exception(data):
+    """Validates a JavaException value
+
+    :arg dict data: the JavaException value
+
+    :raises MalformedJavaException: if the structure is malformed in some way
+
+    """
+    try:
+        jsonschema.validate(data, JAVA_EXCEPTION_SCHEMA)
+        return True
+    except jsonschema.ValidationError as exc:
+        raise MalformedJavaException(exc)
+
+
+def sanitize_java_exception(data):
+    """Removes PII from a JavaException value
+
+    :arg dict data: the JavaException value
+
+    :returns: a deep-copied copy of the data with PII redacted
+
+    """
+    data = copy.deepcopy(data)
+    for item in data["exception"]["values"]:
+        stacktrace = item["stacktrace"]
+        if "value" in stacktrace:
+            stacktrace["value"] = "REDACTED"
```

Comment:
This is the exception message. It can contain PII in cases where the exception is a problem with substring or something like that. Then the message contains the string it was working on and that string could be a url.

Reply:
If we ever discover some other part of the JavaException has PII, we'd need to fix the processor code here and then reprocess all those crashes.

On the one hand, that's probably a lot of crashes.

On the other hand, processing Java crashes is really fast since we don't have to run the stackwalker which has to download and parse symbols in order to walk the stack.

I should add a note here about that.